include/matx/transforms/cub.h

```diff
@@ -1540,6 +1540,133 @@ void sort_impl_inner(OutputTensor &a_out, const InputOperator &a,
 #endif
 }
 
+
+/**
+ * Inner function for the public argsort_impl(). argsort_impl() allocates a temporary
+ * tensor that is contiguous, and can be mutated.
+ */
+template <typename OutputIndexTensor, typename InputIndexTensor, typename OutputKeyTensor,  typename InputKeyTensor>
+void sort_pairs_impl_inner(OutputIndexTensor &idx_out, const InputIndexTensor &idx_in,
+          OutputKeyTensor &a_out, const InputKeyTensor &a_in,
+          const SortDirection_t dir,
+          [[maybe_unused]] cudaExecutor exec = 0)
+{
+#ifdef __CUDACC__
+  MATX_NVTX_START("", matx::MATX_NVTX_LOG_API)
+
+  static constexpr int RANK = OutputIndexTensor::Rank();
+  using T1 = typename InputKeyTensor::value_type;
+  using T2 = typename OutputIndexTensor::value_type;
+
+  cudaStream_t stream = exec.getStream();
+  void *d_temp = nullptr;
+  size_t temp_storage_bytes = 0;
+
+  if constexpr (RANK == 1) {  
+    if (dir == SORT_DIR_ASC) {
+      // First call to get size
+      cub::DeviceRadixSort::SortPairs(d_temp, temp_storage_bytes, 
+                                a_in.Data(), a_out.Data(),
+                                idx_in.Data(), idx_out.Data(),
+                                a_in.Size(0), 0, sizeof(T1) * 8, stream);
+      matxAlloc((void **)&d_temp, temp_storage_bytes, MATX_ASYNC_DEVICE_MEMORY,
+                stream);
+                      
+      // Run sort
+      cub::DeviceRadixSort::SortPairs(d_temp, temp_storage_bytes, 
+                                a_in.Data(), a_out.Data(),
+                                idx_in.Data(), idx_out.Data(),
+                                     a_in.Size(0), 0, sizeof(T1) * 8, stream);
+
+      matxFree(d_temp, stream);
+    }
+    else {
+      cub::DeviceRadixSort::SortPairsDescending(d_temp, temp_storage_bytes, 
+                                a_in.Data(), a_out.Data(),
+                                idx_in.Data(), idx_out.Data(),
+                                     a_in.Size(0), 0, sizeof(T1) * 8, stream);
+      
+      // Allocate temporary storage
+      matxAlloc((void **)&d_temp, temp_storage_bytes, MATX_ASYNC_DEVICE_MEMORY,
+                stream);
+                      
+      // Run sort
+      cub::DeviceRadixSort::SortPairsDescending(d_temp, temp_storage_bytes, 
+                                a_in.Data(), a_out.Data(),
+                                idx_in.Data(), idx_out.Data(),
+                                     a_in.Size(0), 0, sizeof(T1) * 8, stream);
+
+      matxFree(d_temp, stream);
+    }
+  }
+  else {
+    // CUB added support for large items and segments in https://github.com/NVIDIA/cccl/pull/3308
+    int64_t _num_segments = 1;
+    for (int i = 0; i < RANK - 1; i++) {
+      _num_segments *= a_in.Size(i);
```

Comment:
You might be able to use `TotalSize(a_in);` for this

Reply:
I did consider that, but we need the product of all dimension sizes _except_ the last one.